### PR TITLE
Fix/challenge ordering issue

### PR DESCRIPTION
**Pull request by din-mukhammed**
### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- 0chain:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | reward_test_timeout  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | fix/challenge-issues-parallel-din  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 1]
```diff
@@ -1,408 +1,154 @@
...
-	var allOpenChallenges []*ChallengeEntity
+func init() {
+	unProcessedChallengeCh = make(chan *ChallengeEntity, 1)
+}
 
-	var downloadElapsed, jsonElapsed time.Duration
+// syncOpenChallenges will request blockchain if it has been challenged and will receive
+// maximum of 50 challenges in each request. It uses request params to request challenges
+// namely blobberID, limit, from(challenges greater than `from` date) and sort(which should)
+// be descending order.
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:41]: why descending, I believe we want to process challenges starting from the earliest up to latest

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 2]
```diff
@@ -1,408 +1,154 @@
...
+				updateChallengeStatus(chalEntity, Cancelled, err.Error())
+				return
+			}
+			chalEntity.ChallengeTiming.CompleteValidation = common.Now()
+
+			chalEntity.StatusCh <- Completed
+			err = <-chalEntity.ErrCh
+			if err != nil {
+				logging.Logger.Error("Challenge commit error", zap.Error(err))
+				updateChallengeStatus(chalEntity, Cancelled, "Completed but cancelled due to error: "+err.Error())
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:141]: will we write twice in chalEntity.StatusCh?

**Review by dabasov — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 3]
```diff
@@ -1,408 +1,154 @@
...
-		params["from"] = strconv.Itoa(lastChallengeTimestamp)
-	}
-	start := time.Now()
+var (
+	unProcessedChallengeCh chan *ChallengeEntity
+)
 
-	var allOpenChallenges []*ChallengeEntity
+func init() {
+	unProcessedChallengeCh = make(chan *ChallengeEntity, 1)
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:35]: is it enough to have buffer so small?

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 4]
```diff
@@ -1,408 +1,154 @@
...
-		}
+			err = chalEntity.LoadValidationTickets(ctx)
+			if err != nil {
+				updateChallengeStatus(chalEntity, Cancelled, err.Error())
+				return
+			}
+			chalEntity.ChallengeTiming.CompleteValidation = common.Now()
+
+			chalEntity.StatusCh <- Completed
+			err = <-chalEntity.ErrCh
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:138]: what if no error is written, will it hang forever?

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 5]
```diff
@@ -1,408 +1,154 @@
...
+			if err != nil {
+				logging.Logger.Error("Challenge commit error", zap.Error(err))
+				updateChallengeStatus(chalEntity, Cancelled, "Completed but cancelled due to error: "+err.Error())
+				return
+			}
+			chalEntity.Status = Committed
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:144]: why is it committed?

File: code/go/0chain.net/blobbercore/challenge/protocol.go [hunk 6]
```diff
@@ -354,111 +289,311 @@ func (cr *ChallengeEntity) LoadValidationTickets(ctx context.Context) error {
...
+//
+// We need to also take care of the fact that there are other processes(write/read marker redeeming) that uses
+// nonceMonitor which will update latest nonce value.
+
+func ProcessChallengeTransactions(ctx context.Context) {
+	dblLinkedMu := &sync.Mutex{}
+	var latestCResp *ChallengeResponse
+	const guideNum = 10
+	guideCh := make(chan struct{}, guideNum)
+	stopToProcessCh := make(chan *ChallengeResponse, 1)
```
Comment (dabasov) [inline code/go/0chain.net/blobbercore/challenge/protocol.go:329]: why do we need buffered channels here?

**Review by dabasov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**